flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/schema/FlinkRelOptTable.scala

```diff
@@ -0,0 +1,265 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan.schema
+
+import org.apache.calcite.plan.{RelOptCluster, RelOptSchema, RelOptTable}
+import org.apache.calcite.rel.`type`.RelDataType
+import org.apache.calcite.prepare.CalcitePrepareImpl
+import java.util.{List => JList}
+
+import com.google.common.collect.ImmutableList
+import org.apache.calcite.adapter.enumerable.EnumerableTableScan
+import org.apache.calcite.linq4j.tree.Expression
+import org.apache.calcite.plan.RelOptTable.ToRelContext
+import org.apache.calcite.prepare.Prepare.PreparingTable
+import org.apache.calcite.rel.`type`.{RelDataTypeField, RelDataTypeFactory}
+import org.apache.calcite.rel._
+import org.apache.calcite.rel.logical.LogicalTableScan
+import org.apache.calcite.runtime.Hook
+import org.apache.calcite.schema.{StreamableTable, TranslatableTable}
+import org.apache.calcite.sql.SqlAccessType
+import org.apache.calcite.sql.validate.{SqlModality, SqlMonotonicity}
+import org.apache.calcite.sql2rel.InitializerContext
+import org.apache.calcite.util.ImmutableBitSet
+import org.apache.flink.table.plan.stats.FlinkStatistic
+
+import scala.collection.JavaConverters._
+
+/**
+  * FlinkRelOptTable wraps a FlinkTable
+  *
+  * @param schema  the [[RelOptSchema]] this table belongs to
+  * @param rowType the type of rows returned by this table
+  * @param names   the identifier for this table. The identifier must be unique with
+  *                respect to the Connection producing this table.
+  * @param table   wrapped flink table
+  */
+
+class FlinkRelOptTable protected(
+    schema: RelOptSchema,
+    rowType: RelDataType,
+    names: JList[String],
+    table: FlinkTable[_]) extends PreparingTable {
+
+  /**
+    * Creates a copy of this Flink RelOptTable with new Flink Table and new row type.
+    *
+    * @param newTable    new flink table
+    * @param typeFactory type factory to create new row type of new flink table
+    * @return The copy of this Flink RelOptTable with new Flink table and new row type
+    */
+  def copy(newTable: FlinkTable[_], typeFactory: RelDataTypeFactory): FlinkRelOptTable = {
+    val newRowType = newTable.getRowType(typeFactory)
+    FlinkRelOptTable.create(schema, newRowType, names, newTable)
+  }
+
+  /**
+    * Extends a table with the given extra fields, which is not supported now.
+    *
+    * @param extendedFields
+    * @return
+    */
+  override def extend(extendedFields: JList[RelDataTypeField]): RelOptTable =
+    throw new UnsupportedOperationException
+
+  /**
+    * Obtains an identifier for this table.
+    *
+    * @return qualified name
+    */
+  override def getQualifiedName: JList[String] = names
+
+
+  /**
+    * Obtains the access type of the table.
+    *
+    * @return all access types including SELECT/UPDATE/INSERT/DELETE
+    */
+  override def getAllowedAccess: SqlAccessType = SqlAccessType.ALL
```

Comment:
I think we should make this depending on the type of `table`. 
With [FLINK-6442](https://issues.apache.org/jira/browse/FLINK-6442) we added support to register `TableSink` and `INSERT INTO sinkTable SELECT`. 

So if `table` is backed by a `DataSet`, `DataStream`, or `TableSource` is supports `SELECT`. If `table` is backed by a `TableSink` it supports `INSERT` (`DELETE` and `UPDATE` might be added later).